src/compiler/checker.ts

```diff
@@ -31965,6 +32005,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                         // Inferences made from return types have lower priority than all other inferences.
                         inferTypes(context.inferences, inferenceSourceType, inferenceTargetType, InferencePriority.ReturnType);
                     }
+                    // if the contextual type is a tuple and it's is coming from a binding pattern
+                    // then we might need to expand it to match the arity of the target type
+                    // this allows us to infer tuples even if the binding pattern is shorter than the target type, like here:
+                    //   declare function fn<T>(t: T) => [T, string[]];
+                    //   const [[x]] = fn(['hi']);
+                    else if (isTupleType(contextualType) && isTupleType(inferenceTargetType)) {

```

Comment:
this doesn't handle the case when `inferenceTargetType` is a union of tuple types but I have no idea how "expansion" towards that shape should work - unless all union members would have the very same shape